service/storage_service.cc

```diff
@@ -1294,40 +1279,58 @@ future<> storage_service::raft_initialize_discovery_leader(const join_node_reque
-
         insert_join_request_mutations.push_back(std::move(enable_features_mutation));
 
-        auto sl_status_mutation = co_await _sys_ks.local().make_service_levels_version_mutation(2, guard);
+        auto sl_status_mutation = co_await _sys_ks.local().make_service_levels_version_mutation(2, write_timestamp);
         insert_join_request_mutations.emplace_back(std::move(sl_status_mutation));
-        
-        insert_join_request_mutations.emplace_back(
-                co_await _sys_ks.local().make_auth_version_mutation(guard.write_timestamp(), db::system_keyspace::auth_version_t::v2));
+
+        insert_join_request_mutations.emplace_back(co_await _sys_ks.local().make_auth_version_mutation(write_timestamp, db::system_keyspace::auth_version_t::v2));
 
         if (!utils::get_local_injector().is_enabled("skip_vb_v2_version_mut")) {
             insert_join_request_mutations.emplace_back(
-                    co_await _sys_ks.local().make_view_builder_version_mutation(guard.write_timestamp(), db::system_keyspace::view_builder_version_t::v2));
+                    co_await _sys_ks.local().make_view_builder_version_mutation(write_timestamp, db::system_keyspace::view_builder_version_t::v2));
         }
 
         topology_change change{std::move(insert_join_request_mutations)};
-        group0_command g0_cmd = _group0->client().prepare_command(std::move(change), guard,
-                "bootstrap: adding myself as the first node to the topology");
-        try {
-            co_await _group0->client().add_entry(std::move(g0_cmd), std::move(guard), _group0_as, raft_timeout{});
-        } catch (group0_concurrent_modification&) {
-            rtlogger.info("bootstrap: concurrent operation is detected, retrying.");
-        }
-    }
+
+        group0_command group0_cmd{.change{std::move(change)},
+                .history_append{db::system_keyspace::make_group0_history_state_id_mutation(new_group0_state_id,
+                        gc_clock::duration{std::chrono::duration_cast<gc_clock::duration>(std::chrono::weeks{1})},
+                        "bootstrap: adding myself as the first node to the topology")},
+                .prev_state_id{utils::UUID{}},
+                .new_state_id{new_group0_state_id},
+                .creator_addr{_sys_ks.local().local_db().get_token_metadata().get_topology().my_address()},
+                .creator_id{params.host_id}};
+
+        SCYLLA_ASSERT(std::holds_alternative<topology_change>(group0_cmd.change));
+        auto chng = std::get<topology_change>(group0_cmd.change);
+        co_await write_topology_and_other_mutations_to_database(_qp.proxy(), group0_cmd.creator_addr, std::move(chng.mutations));
+        canonical_mutation& history = group0_cmd.history_append;
+        co_await _migration_manager.local().get_storage_proxy().mutate_locally({history.to_mutation(_db.local().find_schema(history.column_family_id()))}, nullptr);
+}
+
+future<> storage_service::raft_initialize_upgrade_state() {
```

Comment:
This function is called `raft_initialize_upgrade_state` but always initializes the upgrade state to `done`. This is not right. We could rename it to something like `initialize_done_topology_upgrade_state`.